### ISSUE TITLE
Add support for `NATURAL JOIN` syntax

### ISSUE THREAD

**Issue by mdesmet** [trinodb/trino#14455]
Currently NATURAL JOIN is not supported in Trino

https://github.com/trinodb/trino/blob/f09633641822f659c1a9e3ddbefa3c5937dffcdc/core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java#L2843

The NATURAL JOIN is used in ETL frameworks to implement deduplication.

For example see this dbt snippet:

```
{%- macro default__deduplicate(relation, partition_by, order_by) -%}

    with row_numbered as (
        select
            _inner.*,
            row_number() over (
                partition by {{ partition_by }}
                order by {{ order_by }}
            ) as rn
        from {{ relation }} as _inner
    )

    select
        distinct data.*
    from {{ relation }} as data
    {#
    -- Not all DBs will support natural joins but the ones that do include:
    -- Oracle, MySQL, SQLite, Redshift, Teradata, Materialize, Databricks
    -- Apache Spark, SingleStore, Vertica
    -- Those that do not appear to support natural joins include:
    -- SQLServer, Trino, Presto, Rockset, Athena
    #}
    natural join row_numbered
    where row_numbered.rn = 1

{%- endmacro -%}
```

I think the natural join could be implemented through rewriting the query with a ON expression on all common columns.

**Comment by findepi** [comment]
Yes, but there is a code snippet that looks as attempting to show why and where it's beneficial over what we support today.

**Comment by martint** [comment]
> where it's beneficial over what we support today.

Other than for language completeness, I don't think there's much benefit to natural joins. In general, they should be avoided, as they can inadvertently start producing unexpected results if, for instance, a new column is added to one of the underlying tables.

The only case I can think of where it makes sense is if you're joining the table to itself, as in the query above.

**Comment by hashhar** [comment]
For dedupe specifically the simplest impl is:

```sql
SELECT * FROM table EXCEPT ALL SELECT DISTINCT * FROM table;
```

To be honest I've never had a case where a NATURAL JOIN couldn't be converted to something more appropriate to task at hand. The only motivation to add it (in my opinion) would be for SQL spec coverage.

**Comment by findepi** [comment]
In the case of the query above, why do we need a Join?
Can't we just do a window and then filter, and then distinct?
